lib/src/default_index/revset_engine.rs

```diff
@@ -900,13 +973,19 @@ impl<'index> EvaluationContext<'index> {
                 let set2 = self.evaluate(expression2)?;
                 Ok(Box::new(DifferenceRevset { set1, set2 }))
             }
+            ResolvedExpression::Extension(extension) => Ok(Box::new(ExternalRevset {
+                revset: extension.evaluate(&|expr| {
```

Comment:
The extension expression will probably need access to the concrete index type (and sometimes repo.) I have no idea how that can be abstracted, but the current implementation appears not supporting that. And I guess some extension expressions are backend specific, and the default revset engine will just error out.

For filter expressions (e.g. select commits by extra metadata), maybe `RevsetFilterPredicate` can have an extension point instead. However, a simple `Extension(Box<dyn Fn(..)>)` wouldn't be okay for the google backend, I suppose.

Reply:
Perhaps, both expression and revset engine need hook point?

* Expression carries an opaque value object which the engine can recognize. (If we don't care type safety, it could be just `Extension(String)`.)
* Default revset engine will just error out. (Alternatively, the extension expression object could provide a fallback function like `StringPattern::matches()`.)
* Google's revset engine serves the role of engine hook point, and it knows how to evaluate the extension expression.

I don't know what kind of extended expression is needed, but it might be good enough to add hook point to `RevsetFilterPredicate`.

Reply:
> The extension expression will probably need access to the concrete index type (and sometimes repo.)

I don't know if this is true generally; at the very least, it's trivial to make the `CompositeIndex<'_>` available to `evaluate()` but I agree I don't know how it's feasible to make the concrete index type accessible.  The repo is also not accessible here at all, so that would be a larger change.

> Perhaps, both expression and revset engine need hook point?  [bullets...]

So all of this is supported by this PR.  The revset engine hook point already exists, that's [Index::evaluate_revset](https://github.com/martinvonz/jj/blob/78edb615a250b8422f4871dfaeeb775d02f12f2d/lib/src/index.rs#L93-L97); Google overrides this internally and doesn't use the default index.  The intention here is that simpler extensions can define their own revset functions in Rust that interop with the default index, while more advanced ones only interop with the parse and resolve steps and always return an Err() on `self.evaluate()` because evaluation is handled by the custom Index implementation.

I'm not opposed to adding an extension variant to RevsetFilterPredicate to fill out this support, but you're correct that it wouldn't be sufficient for Google's implementation (we run custom db queries on the server to handle things like `author()` and `files()` filters on the extremely large set of immutable commits).

Reply:
I had also had something like Yuya's suggestion in mind, plus making it possible to customize the `SymbolResolver` (for cl/123 syntax at Google).

Would that work for our revset engine at Google? I'm not sure if you're saying that it would not or if you had just not considered it. It seems like a more focused extension point, so I would prefer doing it that way if it's possible.